cuML/src/decisiontree/decisiontree.cu

```diff
@@ -0,0 +1,257 @@
+		{
+
+			// Bootstrap columns
+			std::vector<int> colselector(dinfo.Ncols);
+			std::iota(colselector.begin(), colselector.end(), 0);
+			std::random_shuffle(colselector.begin(), colselector.end());
+			colselector.resize((int)(colper * dinfo.Ncols ));
+
+			CUDA_CHECK(cudaHostRegister(colselector.data(), sizeof(int) * colselector.size(), cudaHostRegisterDefault));
+			// Copy sampled column IDs to device memory
+			CUDA_CHECK(cudaMemcpy(tempmem[0]->d_colids, colselector.data(), sizeof(int) * colselector.size(), cudaMemcpyHostToDevice));
+
+			// Optimize ginibefore; no need to compute except for root.
+			if (depth == 0) {
+				int *labelptr = tempmem[0]->sampledlabels;
+				get_sampled_labels(labels, labelptr, rowids, n_sampled_rows);
+				gini(labelptr, n_sampled_rows, tempmem[0], split_info[0], n_unique_labels);
+			}
+
+			int current_nbins = (n_sampled_rows < nbins) ? n_sampled_rows : nbins;
+			best_split_all_cols(data, rowids, labels, current_nbins, n_sampled_rows, n_unique_labels, dinfo.NLocalrows, colselector, tempmem[0], &split_info[0], ques, gain, split_algo);
+
+			//Unregister
+			CUDA_CHECK(cudaHostUnregister(colselector.data()));
+		}
+
+		template<typename T>
+		void DecisionTreeClassifier<T>::split_branch(T *data, GiniQuestion<T> & ques, const int n_sampled_rows, int& nrowsleft, int& nrowsright, unsigned int* rowids)
+		{
+			T *temp_data = tempmem[0]->temp_data;
+			T *sampledcolumn = &temp_data[n_sampled_rows * ques.bootstrapped_column];
+			make_split(sampledcolumn, ques, n_sampled_rows, nrowsleft, nrowsright, rowids, split_algo, tempmem[0]);
+		}
+
+		template<typename T>
+		int DecisionTreeClassifier<T>::classify(const T * row, TreeNode<T> * node, bool verbose) {
+			Question<T> q = node->question;
+			if (node->left && (row[q.column] <= q.value)) {
+				if (verbose)
+					std::cout << "Classifying Left @ node w/ column " << q.column << " and value " << q.value << std::endl;
+				return classify(row, node->left, verbose);
+			} else if (node->right && (row[q.column] > q.value)) {
+				if (verbose)
+					std::cout << "Classifying Right @ node w/ column " << q.column << " and value " << q.value << std::endl;
+				return classify(row, node->right, verbose);
+			} else {
+				if (verbose)
+					std::cout << "Leaf node. Predicting " << node->class_predict << std::endl;
+				return node->class_predict;
+			}
+		}
+
+		template<typename T>
+		void DecisionTreeClassifier<T>::print_node(const std::string& prefix, TreeNode<T>* node, bool isLeft)
+		{
+			if (node != NULL) {
+				std::cout << prefix;
+
+				std::cout << (isLeft ? "├" : "└" );
+
+				// print the value of the node
+				std::cout << node << std::endl;
+
+				// enter the next tree level - left and right branch
+				print_node( prefix + (isLeft ? "│   " : "    "), node->left, true);
+				print_node( prefix + (isLeft ? "│   " : "    "), node->right, false);
+			}
+		}
+
+		//Class specializations
+		template class DecisionTreeClassifier<float>;
```

Comment:
While we already have the great work done to pave the path for #456, and while there are already algorithms in the codebase providing a stateful class-based API, I am hesitant to completely remove the stateful API altogether.

Since we have agreed that all algorithmss provide stateless API at the very least, @myrtw, I would recommend a design similar to what has been done in UMAP, where the statefull (class-based) API is composed of the stateless functions. (Note that the discussion for #456 happened after UMAP was delivered for 0.6 and so the stateless functions still need to be moved to the ML namspace). 

The idea behind such a design is to keep a convenience API that gives users the simplicity and familiar look & feel as sklearn, while still providing, underneath, the stateless functions that enable C- & FORTRAN-based APIs. 

Since we already provide these stateful convenience classes for other algorithms in our public API, I propose we allow @myrtw to do the same here, following the design outlined above, and move all of the stateful convenience classes to a common namespace and/or directory in a follow-in PR, perhaps for 0.7/0.8.

Reply:
I've exposed stateless API functions both for the decision tree and the randomforest. Let me know if you require a further code reorg to be in par w/ other algorithms.